jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResourceFactory.java

```diff
@@ -62,18 +90,39 @@ public void setUseCaches(boolean useCaches)
     @Override
     public Resource newResource(final URI uri)
     {
-        try
-        {
-            return new URLResource(uri, this.connectTimeout, this.useCaches);
-        }
-        catch (MalformedURLException e)
+        URLResource urlResource;
+        try (AutoLock lock = mapLock.lock())
         {
-            throw new RuntimeException("Bad URI: " + uri, e);
+            urlResource = urlResourceMap.get(uri);
+            if (urlResource == null)
+            {
+                try
+                {
+                    urlResource = new URLResource(uri, this.connectTimeout, this.useCaches);
+                    cleanupReferences();
+                    urlResourceMap.put(uri, urlResource);
+                }
+                catch (MalformedURLException e)
+                {
+                    throw new RuntimeException("Bad URI: " + uri, e);
+                }
+            }
         }
+        return urlResource;
     }
 
-    private static class URLResource extends Resource
+    private void cleanupReferences()
     {
+        // TODO
```

Comment:
We should probably do the cleanup before 12.0.0, but perhaps not for beta3.
I was thinking it should be like the JVM's LocaleObjectCache, in that the factory should keep a list of soft references to all resources given out and that we always clean the stale objects on any new resource (which will call close) and then if the factory is closed, we should close all remaining resources, even if still referenced.